hax/test/integration/test_motr.py

```diff
@@ -388,6 +417,134 @@ def my_get(key: str, recurse: bool = False):
+                ('m0conf/nodes/0x6e00000000000001:0x3/processes'
+                 '/0x7200000000000001:0xa/services/0x7300000000000001:0xc',
+                 json.dumps({"name": "ios", "state": "failed"}))]
+            ]
+        elif (key == 'm0conf/nodes/0x6e00000000000001:0x3/processes'
+              '/0x7200000000000001:0x15' and recurse):
+            return [new_kv(k, v) for k, v in [(
+                'm0conf/nodes/0x6e00000000000001:0x3/processes'
+                '/0x7200000000000001:0x15',
+                json.dumps({"name": "m0_server", "state": "failed"})),
+                ('m0conf/nodes/0x6e00000000000001:0x3/processes'
+                 '/0x7200000000000001:0x15/services/0x7300000000000001'
+                 ':0x17',
+                 json.dumps({"name": "ios", "state": "failed"}))]
+            ]
+        elif (key == 'm0conf/nodes/0x6e00000000000001:0x3/processes'
+              '/0x7200000000000001:0x15'):
+            return new_kv('m0conf/nodes/0x6e00000000000001:0x3/processes'
+                          '/0x7200000000000001:0x15',
+                          json.dumps({"name": "m0_server",
+                                      "state": "failed"}))
+        elif (key == 'm0conf/nodes/0x6e00000000000001:0x3/processes'
+              '/0x7200000000000001:0xa'):
+            return new_kv('m0conf/nodes/0x6e00000000000001:0x3/processes'
+                          '/0x7200000000000001:0xa',
+                          json.dumps({"name": "m0_server", "state": "failed"}))
+        elif key == 'm0conf/nodes/localhost/processes/7/services/rms':
+            return new_kv('m0conf/nodes/localhost/processes/7/services/rms',
+                          '17')
+        elif key == 'm0conf/nodes/0x6e00000000000001:0x3':
+            return new_kv('m0conf/nodes/0x6e00000000000001:0x3',
+                          json.dumps({"name": "localhost",
+                                      "state": "M0_NC_UNKNOWN"}))
+        raise RuntimeError(f'Unexpected call: key={key}, recurse={recurse}')
+
+    mocker.patch.object(consul_util.kv, 'kv_get', side_effect=my_get)
+    mocker.patch.object(consul_util.kv, 'kv_put', return_value=0)
+    mocker.patch.object(consul_util, 'update_drive_state')
+    mocker.patch.object(consul_util,
+                        'get_node_fid',
+                        return_value=Fid(0x6e00000000000001, 0x3))
+    mocker.patch.object(consul_util,
+                        'get_node_encl_fid',
+                        return_value=Fid(0x6500000000000001, 0x4))
+
+    # import pudb.remote
+    # pudb.remote.set_trace(term_size=(130, 50), port=9998)
+    motr.broadcast_ha_states([HAState(fid=Fid(0x7200000000000001, 0x15),
+                                      status=ServiceHealth.FAILED)])
+
+    assert not consul_util.update_drive_state.called, \
+        'The drive state should not be updated when MKFS stops'
+
+    traces = motr._ffi.traces
+    assert AssertionPlan(
```

Comment:
@knekrasov, This whole mechanism is amazing! using wrapper function to capture the function args and using them for verification. We have a similar mechanism in Motr which is used to trace function calls. I think we can further extend this tracing mechanism to trace entry and exit of a function. I think we should take it out of the `test` dir into main hax sources.
Such mechanism will help us trace the performance issues like client slow start. What do you think?

Reply:
@mssawant There are two aspects:

1. For 'normal' Pythonic code things like standard `Mock` will work better (one can check whether the function was called, the returned value can either be mocked or a real function can be invoked, but the invocation can still [be spied on](https://wesmckinney.com/blog/spying-with-python-mocks/)). The solution is fine given it is something that works out of the box.
2. I had to implement this `@trace_call` and `AssertionPlan` mini framework because HaxFFI is very hard to mock using standard methods (reason: it is stuffed with ctypes-based functions). And one more item - I wanted to be able to 'query' for HaxFFI calls and see the sequence of the invocations (given it is not a unit test, the amount of calls that can happen is rather large and we are interested in the order); this functionality is not provided by standard Mock.

So long story short, I'm glad you liked the approach and will be happy to improve it. But I'm not sure that we'll want to use it everywhere 😁